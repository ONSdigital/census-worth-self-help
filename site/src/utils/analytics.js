export default class Analytics {
  static trackEvent(category, action, name = "", value = "") {
    if (window._paq) {
      window._paq.push(["trackEvent", category, action, name, value])
    } else {
      console.debug(`Track event not set : ${category} - ${action} - ${name} - ${value}`)
    }
  }

  static trackSiteSearch(keyword, category, searchCount) {
    if (window._paq) {
      window._paq.push(["trackSiteSearch", keyword, category, searchCount])
    } else {
      console.debug(`Track site search : ${keyword} - ${category} - ${searchCount}`)
    }
  }

  static setPageType(pageType) {
    if (typeof window !== `undefined` && window._paq) {
      window._paq.push(["setCustomDimension", 1, pageType])
    }
  }
}