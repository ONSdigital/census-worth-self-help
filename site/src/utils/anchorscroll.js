export default (history, timeout = 1000) => {
  let observer
  let timeoutId

  const reset = () => {
    if (timeoutId) {
      clearTimeout(timeoutId)
      timeoutId = null
    }

    if (observer) {
      observer.disconnect()
    }
  }

  const convertTextToLink = text => {
    return text.toLowerCase().replace(/\s/g, "-")
  }

  const createScrollToElement = hash => {
    return () => {
      let headers = []
      let tags = ["h1", "h2", "h3"]

      tags.forEach(tag => {
        headers = headers.concat(Array.from(document.getElementsByTagName(tag)))
      })

      let element = headers.find(
        header => convertTextToLink(header.textContent) === hash
      )

      if (element) {
        element.scrollIntoView()
        reset()
        return true
      }

      return false
    }
  }

  function scroll(hash) {
    if (typeof hash !== "string") {
      return
    }

    const anchorText = hash.slice(1).toLowerCase()

    if (!anchorText) {
      return
    }

    const scrollToElement = createScrollToElement(anchorText)

    setTimeout(() => {
      if (scrollToElement()) {
        return
      }

      if (window.MutationObserver) {
        observer = new MutationObserver(scrollToElement)

        observer.observe(document, {
          attributes: true,
          childList: true,
          subtree: true
        })
      }

      timeoutId = setTimeout(reset, timeout)
    })
  }

  // listen for changes
  history.listen(({ location, action }) => {
    if (timeoutId) {
      reset()
    }

    if (action !== "POP") {
      return
    }

    if (location.hash) {
      scroll(location.hash)
    }
  })

  // also trigger on load
  if (history.location.hash) {
    scroll(history.location.hash)
  }
}
