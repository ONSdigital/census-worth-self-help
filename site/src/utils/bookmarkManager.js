import Feedback from "./feedback"

export default class BookmarkManager {
  constructor() {
    try {
      this.localStorage = localStorage
      this.bookmarks = JSON.parse(this.localStorage.getItem("bookmarks"))
    } catch (exception) {
      this.localStorage = false
    }

    if (!this.bookmarks) {
      this.bookmarks = []
    }
  }

  getTopBookmarks() {
    return this.bookmarks
  }

  isPageBookmarked(page) {
    return this.bookmarks.includes(page)
  }

  bookmarkPage(page) {
    if (!this.isPageBookmarked(page)) {
      this.bookmarks.push(page)
    }
    this.localStorage &&
      this.localStorage.setItem("bookmarks", JSON.stringify(this.bookmarks))
  }

  unBookmarkPage(page) {
    this.bookmarks = this.bookmarks.filter(function(bookmark) {
      return bookmark !== page
    })

    this.localStorage &&
      this.localStorage.setItem("bookmarks", JSON.stringify(this.bookmarks))
  }

  addBookmarkClickEventToEdges(edges) {
    edges.forEach(
      edge => (edge.node.clickFunction = Feedback.bookmarkClickEvent)
    )
  }
}
