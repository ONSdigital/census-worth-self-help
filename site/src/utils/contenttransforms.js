var sanitizeHtml = require("sanitize-html")

export const transformSources = htmlString => {
  let videoPath = process.env.GATSBY_ASSETS_PATH
    ? process.env.GATSBY_ASSETS_PATH
    : ""
  if (!htmlString) {
    return ""
  }
  return htmlString.replace(/{{TARGET_ASSETS_SRC}}/g, videoPath)
}

export const htmlSanitize = htmlString => {
  return sanitizeHtml(htmlString, {
    allowedTags: sanitizeHtml.defaults.allowedTags.filter(function(tag) {
      return tag !== "iframe"
    }).concat([
      "video",
      "audio",
      "img",
      "source",
      "h1",
      "h2"
    ]),
    allowedAttributes: {
      a: ["href", "name", "target"],
      img: ["src", "title"],
      source: ["src", "type"],
      video: ["controls"],
      audio: ["controls"]
    }
  })
}
