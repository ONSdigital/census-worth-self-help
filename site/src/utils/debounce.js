import _ from "lodash"

// debounce with default delay, or return sync function if delay exactly zero (for unit tests)
export default (fn, delay) => {
  return (delay === 0)
    ? fn
    : _.debounce((...args) => fn(...args), delay || 1000)
}