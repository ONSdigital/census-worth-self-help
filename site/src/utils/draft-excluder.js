const READY_FOR_PRODUCTION = "Ready for Live Site"

// Articles in draft should not be deployed to production. An article is draft if the draftreason for the article
// is anything other than "Ready for production" - see ONS-359.
//
// We also use this Draft Excluder concept to not show articles that are for the call centre only.
//
// Initially this was done using the draft tag. This technique might be removed in the future.
exports.excludeDraftArticle = node => {
  return Boolean(
    process.env.EXCLUDE_DRAFTS && (
      (
        node.frontmatter.tags &&
        node.frontmatter.tags.includes("draft")
      ) || (
        node.frontmatter.draftreason &&
        node.frontmatter.draftreason !== READY_FOR_PRODUCTION
      ) || node.frontmatter.cconly
    )
  )
}
