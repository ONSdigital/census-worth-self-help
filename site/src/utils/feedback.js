export default class Feedback {
  static bookmarkClickEvent(article) {
    Feedback.feedbackEvent("click-event", "bookmark", article)
  }
  static articleWasBookmarked(article) {
    Feedback.feedbackEvent("article-was-bookmarked", "bookmarked", article, 1)
  }
  static articleWasUnBookmarked(article) {
    Feedback.feedbackEvent("article-was-bookmarked", "unbookmarked", article, 1)
  }
  static articleIsUseful(article, reason) {
    Feedback.feedbackEvent(
        "article-feedback-review",
        "POSITIVE REVIEW: " + article,
        reason
    )
    Feedback.feedbackEvent("article-feedback-rating", "rating", article, 1)
  }
  static articleIsNotUseful(article, reason) {
    Feedback.feedbackEvent(
      "article-feedback-review",
      "REVIEW: " + article,
      reason
    )
    Feedback.feedbackEvent("article-feedback-rating", "rating", article, -1)
  }
  static feedbackEvent(category, action, name = "", value = "") {
    if (window._paq) {
      window._paq.push(["trackEvent", category, action, name, value])
    }
  }
}
