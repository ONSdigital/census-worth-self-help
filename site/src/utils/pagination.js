export class PaginationObject {
  constructor(perPage = 5) {
    this.perPage = perPage
    this._offset = 0
  }

  get page() {
    return ~~(this._offset / this._perPage)
  }

  get perPage() {
    return this._perPage
  }

  get offset() {
    return this._offset
  }

  set perPage(perPage) {
    this._perPage = perPage
  }

  goToPage(page) {
    this._offset = this._perPage * page
  }

  filterResults(results) {
    return results.slice(this._offset, this._offset + this._perPage)
  }

  capOffset(total) {
    if (this._offset > total) {
      this._offset = total
    }
  }
}
