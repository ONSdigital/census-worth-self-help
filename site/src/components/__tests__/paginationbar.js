import React from "react"
import renderer from "react-test-renderer"
import PaginationBar from "../paginationbar"
import { PaginationObject } from "../../utils/pagination"
import { render } from "react-testing-library"

describe("PaginationBar", () => {
  const paginationObject = new PaginationObject()

  it("renders correctly", () => {
    const tree = renderer
      .create(<PaginationBar
            total={10}
            paginationObject={paginationObject}
            clickFunction={ () => {} }
            onPageCount={5}
          />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  /* For pages at beginning / middle / end, check that we:
      correctly list the number of elements showing
      have correct back / forward buttons showing
      have correct page numbers showing.
  */
  it("navigation on page 0", () => {
  	paginationObject.goToPage(0)

	const { getByTestId, queryByTestId } = render(<PaginationBar
            total={88}
            paginationObject={paginationObject}
            clickFunction={ () => {} }
            onPageCount={5}
          />)
    expect(getByTestId("pagination-element-showing")).toHaveTextContent("Showing 1-5 of 88")
    expect(queryByTestId('pagination-first-button')).toBeNull()
    expect(queryByTestId('pagination-back-button')).toBeNull()

    expect(queryByTestId('pagination-0-button')).toBeNull()
    expect(getByTestId('pagination-1-button')).toBeDefined()
    expect(getByTestId('pagination-2-button')).toBeDefined()
    expect(queryByTestId('pagination-3-button')).toBeNull()

    expect(getByTestId('pagination-next-button')).toBeDefined()
    expect(getByTestId('pagination-last-button')).toBeDefined()
  })

  it("navigation on middle page", () => {
  	paginationObject.goToPage(6)

	const { getByTestId, queryByTestId } = render(<PaginationBar
            total={88}
            paginationObject={paginationObject}
            clickFunction={ () => {} }
            onPageCount={5}
          />)
    expect(getByTestId("pagination-element-showing")).toHaveTextContent("Showing 31-35 of 88")
    expect(getByTestId('pagination-first-button')).toBeDefined()
    expect(getByTestId('pagination-back-button')).toBeDefined()

    expect(queryByTestId('pagination-5-button')).toBeNull()
    expect(getByTestId('pagination-6-button')).toBeDefined()
    expect(getByTestId('pagination-7-button')).toBeDefined()
    expect(getByTestId('pagination-8-button')).toBeDefined()
    expect(queryByTestId('pagination-9-button')).toBeNull()

    expect(getByTestId('pagination-next-button')).toBeDefined()
    expect(getByTestId('pagination-last-button')).toBeDefined()
  })

  it("navigation on last page", () => {
  	paginationObject.goToPage(17)

	const { getByTestId, queryByTestId } = render(<PaginationBar
            total={88}
            paginationObject={paginationObject}
            clickFunction={ () => {} }
            onPageCount={3}
          />)
    expect(getByTestId("pagination-element-showing")).toHaveTextContent("Showing 86-88 of 88")
    expect(getByTestId('pagination-first-button')).toBeDefined()
    expect(getByTestId('pagination-back-button')).toBeDefined()

    expect(queryByTestId('pagination-16-button')).toBeNull()
    expect(getByTestId('pagination-17-button')).toBeDefined()
    expect(getByTestId('pagination-18-button')).toBeDefined()
    expect(queryByTestId('pagination-19-button')).toBeNull()

    expect(queryByTestId('pagination-next-button')).toBeNull()
    expect(queryByTestId('pagination-last-button')).toBeNull()
  })
})