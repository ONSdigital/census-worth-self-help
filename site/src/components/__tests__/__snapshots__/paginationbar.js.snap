// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginationBar renders correctly 1`] = `
<div
  css={
    Object {
      "map": undefined,
      "name": "d1qmp1",
      "next": undefined,
      "styles": "
        padding: 15px 0px;
      ",
    }
  }
>
  <div
    className="Button-subhead-Style"
    css={
      Object {
        "map": undefined,
        "name": "1dmo2ca",
        "next": undefined,
        "styles": "
          display: flex;
          
    padding: 5px 15px;
  
        ",
      }
    }
  >
    <div
      css={
        Object {
          "map": undefined,
          "name": "1k0aln6",
          "next": undefined,
          "styles": "
            flex-grow: 1;
          ",
        }
      }
      data-testid="pagination-element-showing"
    >
      Showing 
      1
      -
      5
       of 
      10
    </div>
    <div>
      Show 
      5
       per page
    </div>
  </div>
  <div
    css={
      Object {
        "map": undefined,
        "name": "1sf84gv",
        "next": undefined,
        "styles": "
          display: flex;
          justify-content: center;
        ",
      }
    }
  >
    <div
      className="Article-Title-Style clickable"
      css={
        Object {
          "map": undefined,
          "name": "1j335gw",
          "next": undefined,
          "styles": "
        padding: 0px 5px;
        text-decoration: underline;
      ",
        }
      }
      data-testid="pagination-1-button"
      onClick={[Function]}
    >
      1
    </div>
    <div
      className="Article-Title-Style clickable"
      css={
        Object {
          "map": undefined,
          "name": "tvdd9t",
          "next": undefined,
          "styles": "
        padding: 0px 5px;
        text-decoration: none;
      ",
        }
      }
      data-testid="pagination-2-button"
      onClick={[Function]}
    >
      2
    </div>
    <div
      className="Article-Title-Style clickable"
      css={
        Object {
          "map": undefined,
          "name": "tvdd9t",
          "next": undefined,
          "styles": "
        padding: 0px 5px;
        text-decoration: none;
      ",
        }
      }
      data-testid="pagination-next-button"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angle-right fa-w-8 "
        data-icon="angle-right"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 256 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
    <div
      className="Article-Title-Style clickable"
      css={
        Object {
          "map": undefined,
          "name": "tvdd9t",
          "next": undefined,
          "styles": "
        padding: 0px 5px;
        text-decoration: none;
      ",
        }
      }
      data-testid="pagination-last-button"
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-angle-double-right fa-w-14 "
        data-icon="angle-double-right"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 448 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34zm192-34l-136-136c-9.4-9.4-24.6-9.4-33.9 0l-22.6 22.6c-9.4 9.4-9.4 24.6 0 33.9l96.4 96.4-96.4 96.4c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l136-136c9.4-9.2 9.4-24.4 0-33.8z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </div>
  </div>
</div>
`;
